org.mixedrealitytoolkit.uxcore/Experimental/NonNativeKeyboard/NonNativeKeyTouchAdapter.cs

```diff
@@ -0,0 +1,145 @@
+            animatedPosition = defaultPosition + new Vector3(0, 0, AnimationMovementDelta);
+        }
+
+        private void OnEnable()
+        {
+            transform.localPosition = defaultPosition;
+            lastClickTime = Time.time;
+            if (isInitialized)
+            {
+                 buttonCollider.center = buttonColliderDefaultCenter;
+            }
+
+            Initialize();
+        }
+
+        private void Initialize()
+        {
+            if (isInitialized)
+            {
+                return;
+            }
+            isInitialized = true;
+
+            var rectTransform = GetComponent<RectTransform>();
+            buttonCollider = gameObject.EnsureComponent<BoxCollider>();
+            var size = new Vector3(
+                rectTransform.rect.size.x - ColliderMargin,
+                rectTransform.rect.size.y - ColliderMargin, 
+                ColliderThickness);
+            buttonCollider.size = size;
+            buttonColliderDefaultCenter = new Vector3((size.x + ColliderMargin) / 2.0f,
+                (-size.y - ColliderMargin) / 2.0f, ColliderZDelta);
+            buttonCollider.center = buttonColliderDefaultCenter;
+            
+            image = GetComponent<Graphic>();
+            defaultImageColor = image.color;
+            var button = GetComponent<Button>();
+            
+            interactable = gameObject.EnsureComponent<StatefulInteractable>();
+            interactable.firstSelectEntered.AddListener(selectArgs =>
+            {
+                if (selectArgs.interactorObject is not PokeInteractor ||
+                    Time.time - lastClickTime < ReClickDelayTime)
+                {
+                    return;
+                }
+
+                button.onClick.Invoke();
+                StartCoroutine(MoveButton(defaultPosition, animatedPosition));
+            });
+            button.interactable = false;
+
+            interactable.lastSelectExited.AddListener(_ =>
+            {
+                StartCoroutine(MoveButton(animatedPosition, defaultPosition));
+            });
+
+            interactable.firstHoverEntered.AddListener(hoverArgs =>
+            {
+                SetColorOnHoverPoke(hoverArgs.interactorObject, button.colors.highlightedColor);
+            });
+            
+            interactable.lastHoverExited.AddListener(hoverArgs =>
+            {
+                SetColorOnHoverPoke(hoverArgs.interactorObject, defaultImageColor);
+            });
+        }
+
+        private void SetColorOnHoverPoke(IXRHoverInteractor interaction, Color color)
+        {
+            if (interaction is PokeInteractor)
+            {
+                image.color = color;
+            }
+        }
+        
+        private void Update()
+        {
+            // fallback for if the user very rapidly moves their finger over the keyboard and
+            // interactable.lastHoverExited does not get fired (or too early)
```

Comment:
Seems like you are working around a bug here.   Is there something else we can fix to prevent this component from "missing" the lastHoverExited event?

Reply:
I think I am working around a bug indeed, but I fear that might be in StatefulInteractable. This typically happens when you bring your finger at 'hover distance' from a button, let's say Q, and then rapidly move your finger over the keyboard to the right. It looks like events are then fired in the wrong order. I changed my code to not use the lastHoverExited at all.